receiver/googlecloudspannerreceiver/internal/filterfactory/filterbymetricbuilder.go

```diff
@@ -0,0 +1,154 @@
+		}
+	}
+
+	return filterByMetric
+}
+
+func (b filterByMetricBuilder) buildFilterByMetricPositiveTotalLimit() (map[string]filter.ItemFilter, error) {
+	filterByMetric := make(map[string]filter.ItemFilter)
+	groupedItems := groupByCardinality(b.config.MetadataItems)
+
+	// Handle metric groups with low cardinality
+	lowCardinalityGroups := groupedItems[false]
+	newTotalLimit, err := b.handleLowCardinalityGroups(lowCardinalityGroups, b.config.TotalLimit, filterByMetric)
+	if err != nil {
+		return nil, err
+	}
+
+	// Handle metric groups with high cardinality
+	highCardinalityGroups := groupedItems[true]
+	newTotalLimit, err = b.handleHighCardinalityGroups(highCardinalityGroups, newTotalLimit, filterByMetric)
+	if err != nil {
+		return nil, err
+	}
+
+	b.logger.Debug("Remaining total limit after cardinality limits calculation",
+		zap.Int("remainingTotalLimit", newTotalLimit))
+
+	return filterByMetric, nil
+}
+
+func (b filterByMetricBuilder) handleLowCardinalityGroups(groups []*metadata.MetricsMetadata, remainingTotalLimit int,
+	filterByMetric map[string]filter.ItemFilter) (int, error) {
+
+	if len(groups) == 0 {
+		return remainingTotalLimit, nil
+	}
+
+	limitPerMetricByTimestamp := b.config.ProjectAmount * b.config.InstanceAmount * b.config.DatabaseAmount
+	// For low cardinality metrics total limit is equal to limit by timestamp
+	totalLimitPerMetric := limitPerMetricByTimestamp
+
+	b.logger.Debug("Calculated cardinality limits for low cardinality metric group",
+		zap.Int("limitPerMetricByTimestamp", limitPerMetricByTimestamp),
+		zap.Int("totalLimitPerMetric", totalLimitPerMetric))
+
+	return b.constructFiltersForGroups(totalLimitPerMetric, limitPerMetricByTimestamp, groups, remainingTotalLimit, filterByMetric)
+}
+
+func (b filterByMetricBuilder) handleHighCardinalityGroups(groups []*metadata.MetricsMetadata, remainingTotalLimit int,
+	filterByMetric map[string]filter.ItemFilter) (int, error) {
+
+	if len(groups) == 0 {
+		return remainingTotalLimit, nil
+	}
+
+	totalLimitPerMetric := remainingTotalLimit / totalAmountOfMetricsInGroups(groups)
+	limitPerMetricByTimestamp := totalLimitPerMetric / defaultMetricDataPointsAmountInPeriod
+
+	b.logger.Debug("Calculated cardinality limits for high cardinality metric group",
+		zap.Int("limitPerMetricByTimestamp", limitPerMetricByTimestamp),
+		zap.Int("totalLimitPerMetric", totalLimitPerMetric))
+
+	if limitPerMetricByTimestamp < 1 {
+		return remainingTotalLimit, errors.New("limit per metric per timestamp for high cardinality metrics is lower than 1")
+	}
+
+	return b.constructFiltersForGroups(totalLimitPerMetric, limitPerMetricByTimestamp, groups, remainingTotalLimit, filterByMetric)
+}
+
+func (b filterByMetricBuilder) constructFiltersForGroups(totalLimitPerMetric int, limitPerMetricByTimestamp int,
+	groups []*metadata.MetricsMetadata, remainingTotalLimit int, filterByMetric map[string]filter.ItemFilter) (int, error) {
+
+	newTotalLimit := remainingTotalLimit
```

Comment:
```suggestion
limit := remaining
```

Reply:
Please keep the variable names concise. 🙏🏽

Reply:
In this case I do not agree, because with such names it is not clear that those variables operate with total limit. Of course I can do it, but I don't see a lot of sense here.

Reply:
Mmmmm, sounds reasonable to me just conscious of variable names that make up multiple words as it is typically go idiomatic.